### ISSUE TITLE
SIGSEGV in krb5asrep, Kerberos 5 AS-REP etype 23 [MD4 HMAC-MD5 RC4]

### ISSUE THREAD

**Issue by claudioandre** [magnumripper/JohnTheRipper#2774]
A list of hashes that causes the failure might be shared later. The offending code:
```
#0  JOHN_RC4 (ctx=ctx@entry=0x7fffffff94a0, len=4294967280, ...) at rc4.c:66
#1  0x000055555576af36 in crypt_all (pcount=0xef7dc7a4ec0ab9e3, salt=<optimized out>) at krb5_asrep_fmt_plug.c:323
```
len **is** wrong.
****
* What happens:
cur_salt->edata2len value is 16
len =  (cur_salt->edata2len - 32) at krb5_asrep_fmt_plug.c:323

**Comment by kholia** [comment]
@claudioandre Please share the hash(es) which cause this crash.

**Comment by claudioandre** [comment]
I guess I can't, contest is still active; but you should have it in your mail box.

**Comment by kholia** [comment]
Here is a problematic hash -> `000372dab3df425ed408e6b105741864$be03c4da216b8787bc0d76b721ae9c4b`.

A simple patch to solve this crash,

``` diff
diff --git a/src/krb5_asrep_fmt_plug.c b/src/krb5_asrep_fmt_plug.c
index dae2006..ded1aa4 100644
--- a/src/krb5_asrep_fmt_plug.c
+++ b/src/krb5_asrep_fmt_plug.c
@@ -157,7 +157,7 @@ static int valid(char *ciphertext, struct fmt_main *self)
        /* assume edata2 following */
        if (((p = strtokm(NULL, "$")) == NULL))
                goto err;
-       if (!ishex(p) && (hexlen(p, &extra) < 64 || extra))
+       if (!ishex(p) || (hexlen(p, &extra) < 64 || extra))
                goto err;
 
        if ((strtokm(NULL, "$") != NULL))
```

Using `&&` instead of `||` does not seem to be correct. The `krb5_tgs_fmt_plug.c` file has a similar bug in it.

While we are at it, should we increase the minimum valid length of `edata2` from its current value of `64`? I am not sure if calling the `RC4` function with `len == 0` is OK. This happens when the hex-length of `edata2` is exactly `64`.

**Comment by kholia** [comment]
This issue is fixed now (PR #2791).
